### ISSUE TITLE
panic when compiling huge .rs of 20834 lines

### ISSUE THREAD

**Issue by crownedgrouse** [rust-lang/rust#48620]
While compiling huge .rs of 20834 lines (cannot copy here)
(a fn with 256 matches using other matches ).

When running  `cargo check`  no error (but a very long compilation)  : 
    ` Finished dev [unoptimized + debuginfo] target(s) in 570.49 secs`

## Meta
rustc 1.24.0 (4d90ac38c 2018-02-12)
binary: rustc
commit-hash: 4d90ac38c0b61bb69470b61ea2cccea0df48d9e5
commit-date: 2018-02-12
host: i686-unknown-linux-gnu
release: 1.24.0
LLVM version: 4.0

Backtrace : 
```
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.24.0 (4d90ac38c 2018-02-12) running on i686-unknown-linux-gnu

note: run with `RUST_BACKTRACE=1` for a backtrace

thread 'rustc' panicked at 'index 125594976 out of range for slice of length 125594956', /checkout/src/libcore/slice/mod.rs:750:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at /checkout/src/libstd/sys_common/backtrace.rs:68
             at /checkout/src/libstd/sys_common/backtrace.rs:57
   2: std::panicking::default_hook::{{closure}}
             at /checkout/src/libstd/panicking.rs:381
   3: std::panicking::default_hook
             at /checkout/src/libstd/panicking.rs:391
   4: std::panicking::rust_panic_with_hook
             at /checkout/src/libstd/panicking.rs:577
   5: std::panicking::begin_panic
             at /checkout/src/libstd/panicking.rs:538
   6: std::panicking::begin_panic_fmt
             at /checkout/src/libstd/panicking.rs:522
   7: rust_begin_unwind
             at /checkout/src/libstd/panicking.rs:498
   8: core::panicking::panic_fmt
             at /checkout/src/libcore/panicking.rs:71
   9: core::slice::slice_index_len_fail
             at /checkout/src/libcore/slice/mod.rs:750
  10: <rustc_mir::dataflow::AllSets<E>>::for_block
  11: <rustc_mir::dataflow::DataflowAnalysis<'a, 'tcx, BD>>::build_sets
  12: rustc_mir::dataflow::do_dataflow
  13: <rustc_mir::transform::elaborate_drops::ElaborateDrops as rustc_mir::transform::MirPass>::run_pass
  14: rustc_mir::transform::optimized_mir::{{closure}}
  15: rustc_mir::transform::optimized_mir
  16: rustc::ty::maps::<impl rustc::ty::maps::queries::optimized_mir<'tcx>>::compute_result
  17: rustc::dep_graph::graph::DepGraph::with_task_impl
  18: rustc_errors::Handler::track_diagnostics
  19: rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::cycle_check
  20: rustc::ty::maps::<impl rustc::ty::maps::queries::optimized_mir<'tcx>>::force
  21: rustc::ty::maps::<impl rustc::ty::maps::queries::optimized_mir<'tcx>>::try_get
  22: rustc::ty::maps::TyCtxtAt::optimized_mir
  23: rustc::ty::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::instance_mir
  24: rustc_mir::monomorphize::collector::collect_items_rec
  25: rustc_mir::monomorphize::collector::collect_items_rec
  26: rustc_mir::monomorphize::collector::collect_items_rec
  27: rustc_mir::monomorphize::collector::collect_crate_mono_items
  28: rustc::util::common::time
  29: rustc_trans::base::collect_and_partition_translation_items
  30: rustc::ty::maps::<impl rustc::ty::maps::queries::collect_and_partition_translation_items<'tcx>>::compute_result
  31: rustc::dep_graph::graph::DepGraph::with_task_impl
  32: rustc_errors::Handler::track_diagnostics
  33: rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::cycle_check
  34: rustc::ty::maps::<impl rustc::ty::maps::queries::collect_and_partition_translation_items<'tcx>>::force
  35: rustc::ty::maps::<impl rustc::ty::maps::queries::collect_and_partition_translation_items<'tcx>>::try_get
  36: rustc::ty::maps::TyCtxtAt::collect_and_partition_translation_items
  37: rustc::ty::maps::<impl rustc::ty::context::TyCtxt<'a, 'tcx, 'lcx>>::collect_and_partition_translation_items
  38: rustc_trans::base::trans_crate
  39: <rustc_trans::LlvmTransCrate as rustc_trans_utils::trans_crate::TransCrate>::trans_crate
  40: rustc::util::common::time
  41: rustc_driver::driver::phase_4_translate_to_llvm
  42: rustc_driver::driver::compile_input::{{closure}}
  43: <std::thread::local::LocalKey<T>>::with
  44: <std::thread::local::LocalKey<T>>::with
  45: rustc::ty::context::TyCtxt::create_and_enter
  46: rustc_driver::driver::compile_input
  47: rustc_driver::run_compiler

error: Could not compile `preach`.

```

**Comment by retep998** [comment]
Could you upload the code to a repository somewhere?

**Comment by crownedgrouse** [comment]
I splitted my big function into 4 smaller ones in individual modules (a .rs file per function), but now no error, but rustc fails : 
```
error: Could not compile `preach`.

Caused by:
  process didn't exit successfully: `rustc --crate-name preach src/main.rs --crate-type bin --emit=dep-info,link -C debuginfo=2 -C metadata=10cd77d00eccd0d1 -C extra-filename=-10cd77d00eccd0d1 --out-dir /home/eric/dev/projects/preach/target/debug/deps -C incremental=/home/eric/dev/projects/preach/target/debug/incremental -L dependency=/home/eric/dev/projects/preach/target/debug/deps --extern env_logger=/home/eric/dev/projects/preach/target/debug/deps/libenv_logger-8b8781143d9dedb1.rlib --extern rand=/home/eric/dev/projects/preach/target/debug/deps/librand-7adbc532e0c445b3.rlib --extern getopts=/home/eric/dev/projects/preach/target/debug/deps/libgetopts-bdad2372c882101d.rlib --extern log=/home/eric/dev/projects/preach/target/debug/deps/liblog-bfec1262ec0fbc2e.rlib` (signal: 11, SIGSEGV: invalid memory reference)

```

**Comment by crownedgrouse** [comment]
I will push my code on github. let you know.

**Comment by crownedgrouse** [comment]
check  https://github.com/crownedgrouse/preach   
but this is my version with splitted functions.

**Comment by crownedgrouse** [comment]
RUST_BACKTRACE=1  does not give more usefull info for the version in the repository.

**Comment by crownedgrouse** [comment]
ping ?

**Comment by crownedgrouse** [comment]
tried  rustc 1.30.0   and now compile OK.  Issue was probably fixed since 1.24.0 .
Keep ticket opened, let you close with reference to the commit that fixed that ?

**Comment by Dylan-DPC** [comment]
Closing this as it complies in 1.30 and there is no mcve
